### DISCUSSION TITLE
Feature request: Support for filter queries as json

### DISCUSSION THREAD

**Discussion by slatian** [meilisearch/product#722] [Feedback & Feature Proposal]
Both the facets endpoint and the filter attribute support filter strings (and filter strings in arrays).

While they are easy to use they have problem that one really has to watch out that everything is serialized correctly. The people working on Apache solr have figured this out quite some time ago and implemented their JSON API.

Meiliserch should have the capability to pass filters as JSON too.

So instead of passing the following [example query](https://www.meilisearch.com/docs/learn/fine_tuning_results/filtering#combining-arrays-and-strings):
```
"(genres = comedy OR genres = horror) AND director != 'Jordan Peele'"
```
or as array:
```json
[["genres = comedy, genres = horror"], "NOT director = 'Jordan Peele'"]
```
(Is the nested array containing only one string intentional here?)

One should be able to generate a filter like this:
```json
[
  [
    {
      "op": "=",
      "field": "genres",
      "value": "comedy"
    },{
      "op": "=",
      "field": "genres",
      "value": "horror"
   }
  ],{
    "op": "=",
    "invert": true,
    "field": "director",
    "value": "Jordan Peele"
  }
]
```

Also note that there should be no *magic* behavior based on the content of the `field` and `value` keys. This way one can focus on the application without worrying that some funny user/API supplied string screws up search queries.

**Comment by macraig** [comment]
Hi @slatian , thanks for taking the time to leave your feedback. I understand that you'd like to remove the need to make sure everything is serialized correctly if possible. Would you mind telling us more about your use case? Would you rate the need for JSON support as critical, important, nice to have? Why?

> (Is the nested array containing only one string intentional here?)

The syntax in the other examples uses different strings, but I believe the example you shared is under "Combining arrays and strings". It's not clear to me either, I'll check with the engineering team and get back to you on this.

**Reply by slatian** [reply]
> Would you mind telling us more about your use case?

I'm currently tinkering with small scale web searching (with the count of indexed pages well within of what meilisearch should be capable of) with a focus on also making site-metadata queryable.

> Would you rate the need for JSON support as critical, important, nice to have? Why?

For me the more important than is JSON/isn't JSON is the serialization part and being sure that data doesn't mix with instructions no matter how malicious the input. (So placeholders would work too, but they are less easy to work with for automated query assembly, also meili has half of the JSON API already in place)

The support for such an API-Property is critical enough that I reevaluated all of the other options out there systematically ruling out search engines that don't support it. (Even if that means a drop in search quality or having to use a historically grown API)

> It's not clear to me either, I'll check with the engineering team and get back to you on this.

Thanks.

**Reply by macraig** [reply]
Hi again!

> (Is the nested array containing only one string intentional here?)

I've confirmed this wasn't intentional, we've already updated the documentation to reflect the correct syntax. Thank you for flagging it!
